sdk/program/src/pubkey.rs

```diff
@@ -379,6 +379,7 @@ impl Pubkey {
             stake::config::id(),
             vote::program::id(),
             secp256k1_program::id(),
+            ed25519_program::id(),
```

Comment:
Oof, so I think this change is a consensus footgun.  We probably need a big warning at the top of `is_native_program_id()` that says, DON'T ADD NEW IDs.

The problem is that the `sol_try_find_program_address()` syscall invokes `create_program_address`, which invokes `is_native_program_id()`.  Since there's no feature gating in `is_native_program_id()`, I could fork nodes off the cluster by invoking `sol_try_find_program_address()` from a program with `ed25519_program::id()` as nodes with/without this line will behave differently.  Time it right and you can take out the entire cluster

Reply:
I'm glad you caught that before we had a disaster on our hands :pray: 

So I've started working on a solution where so`sol_try_find_program_address()` and `sol_create_program_address()` use the invoke context list of native programs to check. This requires `Pubkey::create_program_address()` to not check `Pubkey::is_native_program_id()` (which is unchanged). I haven't quite settled on this yet, and it needs more tests, but would you mind giving it a quick look and see what you think (last commit in PR now)

Reply:
The `create_program_address_program_id_checked()` approach looks good to me.  It'd be helpful to add a function comment to `create_program_address_program_id_checked()` that describes when it's safe to use this function over the original

Reply:
So I've cleaned up the code and it's ready for review. However, I think it would be nice to have a test that checks that `sol_try_find_program_address()` and `sol_create_program_address()` work with the ed25519 program id when the feature is not enabled. Is there an easy way to do this or should I just leave it?

Note that `Pubkey::create_program_address()` is still used by CPI, but since the caller can not be the ed25519 program, I assume this is safe.

Reply:
I think the latest commits look good, @jackcmay can you please carefully review as well: `syscalls.rs` and `pubkey.rs`

Reply:
Yup, on my list for today :-)

Reply:
I don't think the native program id checks in this pr are correct.  `sol_try_find_program_address()` and `sol_create_program_address` is enforcing that the program id is not one of the native programs ids but that is not the same as built-in programs.  For example, ids like `feature::id()` and `sysvar::id()` are not included in the builtins.

I think the existing native program id checks in pubkey are not ideal, it creates another place where we have to add a program id when programs are added and can easily be missed (like it was when I added `compute_budget_program`).  I'm not sure yet what the better solution looks like, at the very least the program id check in the syscalls should include the missing ids in addition to the builtins.  I'd rather not but seems like we need a third list where `is_native_program_id` = `builtins` + non-builtin-ids.

I also think we should make these changes in its own PR so folks can review them on their own.

Reply:
I'm putting together a pr to address this, let me know if you have any thoughts

Reply:
I was thinking we should make an explicit allow-list of native programs for these two functions at vm creation time. This both makes it very explicit and also means the list is empty when all features are enabled, making it performant.

Reply:
Fixed